test/testutil.h

```diff
@@ -141,10 +143,201 @@ __owur int run_tests(const char *test_prog_name);
+    DECLARE_COMPARISON(type, name, opname)                              \
+    DEFINE_COMPARISON(type, name, opname, op, fmt)
+
+#define IMPLEMENT_COMPARISONS(type, name, fmt)                          \
+    DECLARE_COMPARISONS(type, name)                                     \
+    DEFINE_COMPARISONS(type, name, fmt)
+
+/*
+ * Define some comparisons between pairs of various types.
+ * These functions return 1 if the test is true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ *
+ * In each case the functions produced are:
+ *  int test_name_eq(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_ne(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_lt(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_le(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_gt(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_ge(const type t1, const type t2, const char *desc, ...);
+ *
+ * The t1 and t2 arguments are to be compared for equality, inequality,
+ * less than, less than or equal to, greater than and greater than or
+ * equal to respectively.  If the specified condition holds, the functions
+ * return 1.  If the condition does not hold, the functions print a diagnostic
+ * message and return 0.
+ *
+ * The desc argument is a printf format string followed by its arguments and
+ * this is included in the output if the condition being tested for is false.
  */
-int strings_equal(const char *desc, const char *s1, const char *s2);
+DECLARE_COMPARISONS(int, int)
+DECLARE_COMPARISONS(unsigned int, uint)
+DECLARE_COMPARISONS(char, char)
+DECLARE_COMPARISONS(unsigned char, uchar)
+DECLARE_COMPARISONS(long int, lint)
+DECLARE_COMPARISONS(unsigned long int, ulint)
+
+/*
+ * Pointer comparisons against other pointers and null.
+ * These functions return 1 if the test is true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ */
+DECLARE_COMPARISON(void *, ptr, eq)
+DECLARE_COMPARISON(void *, ptr, ne)
+int test_ptr(const void *, const char *, ...);
+int test_ptr_null(const void *, const char *, ...);
+
+/*
+ * Equality tests for strings where NULL is a legitimate value.
+ * These calls return 1 if the two passed strings compare true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ */
+DECLARE_COMPARISON(char *, str, eq)
+DECLARE_COMPARISON(char *, str, ne)
+
+/*
+ * Check a boolean result for being true or false.
+ * They return 1 if the condition is true (i.e. the value is non-zro).
+ * Otherwise, they return 0 and pretty-prints diagnostics using |desc|.
+ */
+int test_true(int, const char *desc, ...);
+int test_false(int, const char *desc, ...);
+
+/*
+ * The following macros provide wrapper calls to the test functions with
+ * a default description that indicates the file and line number of the error.
+ */
+#define TEST_int_eq(a, b)   test_int_eq(a, b, "%s: %d", __FILE__, __LINE__)
```

Comment:
I meant you'd do `TEST_int_eq_msg`same as `TEST_error` - variadic macros where you can, and  fall back to a function (losing the file and line number) otherwise. I _think_ that works?

Re file and line numbers, I don't think they ever hurt. I'd probably have `TEST_error` for errors (starts with ERROR or FAIL as you have it) and `TEST_info` for info (starts with INFO) to make it easy to grep for failure lines - but I don't really have a strong opinion. Whatever feels most natural to you!

Re test_xxx functions, I think so? It's not so important - I only suggested moving the declarations to another header to make it easier to find the public bits you're supposed to use in tests.